[FEATURE] Add the functionality to select and create groups

ubarsaiyan opened the pull request:
Closes #194
Also closes #282 
#### Changes:
Added the functionality to select and create groups. 
Added a horizontal list of selected contacts and a screen to name the group.
Also fixed many bugs and crashes, some unrelated to this functionality. 

#### Video for the change:
https://photos.app.goo.gl/EvZq8wotz3hP6rxk9

ubarsaiyan committed:
Initial working implementation of creating groups

ubarsaiyan committed:
Add the menu FAB button (opens the same thing for now)

ubarsaiyan committed:
Enable selection by default when opening create group from the FAB

ubarsaiyan committed:
Show selected contacts list while creating a group

ubarsaiyan committed:
Create new layout for the group details screen

ubarsaiyan committed:
Disable search on group creation and fix search crash

ubarsaiyan committed:
UI improvements

ubarsaiyan committed:
Fix back button crashes

ubarsaiyan committed:
Fix many crashes and improve code quality and logic

ear-dev commented:
@ubarsaiyan Right now I am just testing, and will do the actual code review after that.... and I will start posting a few things that I see in the meantime..

- Clicking the new_chat fab results in the arrow_fab in the bottom of the contacts page.  That fab should only be for the create group page.
- There is no search on the create group page.  I believe the wireframe includes search?
- The group page still includes the INVITE CONTACTS list, but that is not applicable here?  It should be OTHER USERS that can be discovered through search.
- The group name page does not include any version of the list of users you have selected.  Did we decide to include that list with avatars instead of the grey tabs?  That is included in the wireframe.

ear-dev commented:
- From inside the group name page, if you hit the back button because you forgot to add somebody to the group, the group is gone and you have to start over.  Is there any way to maintain that state until the user backs out of the new group page entirely?

ear-dev commented:
- Clicking on the "new group" text box (next to the fab) results in the New Chat page opening.

ubarsaiyan commented:
1. [FIXED] ~Clicking the new_chat fab results in the arrow_fab in the bottom of the contacts page. Yes! Sorry, I'll fix this.~
2. [ADDED] ~There is no search on the create group page. I have listed it in the TODOs. It's a bit tricky to implement so I have left it for now.~
3. [NOT TO BE DONE] ~The group page still includes the INVITE CONTACTS list, but that is not applicable here? It should be OTHER USERS that can be discovered through search. Oh... I didn't know about this but it makes sense. I'll look into this.~
4. [DONE] ~The group name page does not include any version of the list of users you have selected. I think we agreed to add a list later. Should I include it now only?~
5. [DONE] ~From inside the group name page, if you hit the back button because you forgot to add somebody to the group, the group is gone and you have to start over. Is there any way to maintain that state until the user backs out of the new group page entirely? I tried to retain the state but it was then causing other problems like duplicate selected items being shown. I then rewrote the whole adapter of the recycler view as without it nothing could be done. I'll try again to fix this issue.~
6. [FIXED] ~Clicking on the "new group" text box (next to the fab) results in the New Chat page opening. My bad. Fix incoming.~
7. [FIXED] ~After the user has clicked the "allow permissions" system dialog from the permission item, the fragment is closed and the user ends up back in the chatrooms view. This PR has broken that. Sorry, I'll fix it.~

ear-dev commented on app/src/main/java/chat/rocket/android/chatrooms/ui/ChatRoomsFragment.kt:
```diff
@@ -543,14 +549,83 @@ class ChatRoomsFragment : Fragment(), ChatRoomsView {
+        }
+        new_group_fab_item.setOnClickListener {
+            hideFABMenu()
+            openFragment(ContactsFragment(), "contactsFragment")
```
Need "enableGroups = true" here.

ubarsaiyan replied:
Yup. Done.

ear-dev commented on app/src/main/java/chat/rocket/android/contacts/adapter/ContactsRecyclerViewAdapter.kt:
```diff
@@ -41,11 +45,16 @@ class ContactsRecyclerViewAdapter(
                 val view = parent.inflate(R.layout.item_permissions)
                 PermissionsViewHolder(view)
             }
-            else -> throw IllegalStateException("View type must be either Contact, Header, Invite or Permissions")
```
You don't like "Contact" or "Permissions" in these error messages?

ubarsaiyan replied:
Sorry... Must be caused due to rebasing the new commits

ubarsaiyan committed:
Fix fab in contacts page

ear-dev commented on app/src/main/java/chat/rocket/android/contacts/adapter/ContactsRecyclerViewAdapter.kt:
```diff
@@ -105,38 +120,33 @@ class ContactsRecyclerViewAdapter(
                     }
                 }
             }
-              // Clicking the @username button will open a DM
```
We have been having a debate about whether to allow the whole row or just the @username button to open the DM.  Do you mind leaving this in so that it will be easy to toggle between modes in the case we want to make a change?  Maybe comment it out with a NOTE if you think that's better........  thanks.

ubarsaiyan replied:
Sure

ear-dev commented on app/src/main/java/chat/rocket/android/contacts/ui/ContactsFragment.kt:
```diff
@@ -48,20 +49,29 @@ import chat.rocket.android.util.extensions.ui
+    //FIXME: Remove the blink when a contact is selected again after all the contacts are deselected
+    //TODO: When the group is being selected via long press from the new chat screen, change the title to create group from new chat
+    //TODO: Add support for square avatars during group creation flow
+
```
Your plan is to do all of the above before we merge this?  Just checking......

ubarsaiyan replied:
Not everything I guess... Whatever the team finds important enough to be done before merging

ear-dev commented on app/src/main/java/chat/rocket/android/contacts/ui/ContactsFragment.kt:
```diff
@@ -195,6 +252,9 @@ class ContactsFragment : Fragment(), ContactsView {
         searchView?.onQueryTextListener { queryContacts(it) }
 
+        if (enableGroups)
+            searchItem.isVisible = false
```
We need search to populate OTHER USERS

ubarsaiyan replied:
I disabled this currently because selecting the users via search was causing problems (as the items disappear after the searchview is closed.

ubarsaiyan replied:
Will do. Noted.

ear-dev commented:
"The group page still includes the INVITE CONTACTS list, but that is not applicable here? It should be OTHER USERS that can be discovered through search. Oh... I didn't know about this but it makes sense. I'll look into this." 
- Do you still have the link to the wireframe?  That will currently be our source of truth...

ear-dev commented:
In the case that the user has not given permission to allow VC to access contacts, we added that permissions_item to the contacts page (I believe you saw that).  In the veranda branch, when you use that to allow permissions eventually, after the user has clicked the "allow permissions" system dialog, the fragment is closed and the user ends up back in the chatrooms view.  This PR has broken that.  Can you fix please?  Thanks.

ear-dev commented:
"The group name page does not include any version of the list of users you have selected. I think we agreed to add a list later. Should I include it now only?"
- Ok, I guess that could be up to you.  It's fine to leave that as a task for later, but if you do that, can you create a story in the contacts project to track it?

ear-dev commented on app/src/main/res/values/strings.xml:
```diff
@@ -377,6 +379,8 @@ https://github.com/RocketChat/java-code-styles/blob/master/CODING_STYLE.md#strin
     <string name="msg_widechat_no_chats_to_display">You can get started by searching for people to chat with or by inviting your friends and family to get the app too.</string>
     <string name="msg_widechat_welcome_to_app">Welcome to Viasat Connect.</string>
+    <string name="new_group">New Group</string>
+    <string name="new_contact">New Contact</string>
```
Any chance you can add spanish translations as well?  If that's not possible we can just make a story about it and do it later...

ubarsaiyan replied:
I can google translate and then you can verify perhaps?

ear-dev replied:
Google translate is fine for now.... thanks!

ubarsaiyan replied:
Done

ear-dev reviewed (commented):
This is a really good PR.  Thanks.  I've made just a few minor comments so far, some of which may already be obsolete... :-)

ubarsaiyan commented:
@ear-dev I'm currently doing the showing of the selected contacts list in the group naming screen as I think it's important and may also come in handy while debugging if any user is not added to the group. 
About the OTHER USERS thing, the wireframe https://sketch.cloud/s/8YkrR/ygYKMRA shows "OTHER USERS" written but I didn't give it much thought and assumed it to be the same as INVITE CONTACTS. Can we discuss in the meeting about this?
And about the allow permissions thing, I'm going to fix it :+1:

ear-dev commented:
About OTHER USERS:  In the wireframe, if you click the search Icon, and then click the search bar, you will see how OTHER USERS is designed to work.... FYI

ubarsaiyan committed:
Add group members list in the group naming screen

ubarsaiyan commented:
@ear-dev about the pt 7 [here](https://github.com/WideChat/Rocket.Chat.Android/pull/292#issuecomment-475269291). Are you sure it worked before? Because I don't see any code that I have removed that took the user back to the main screen. EDIT: I have fixed it... If you find your earlier code to be better pls point me to it and I will replace this code.

ubarsaiyan committed:
Fix contacts fragment not destroyed after user grants permission manually

ubarsaiyan committed:
Fix search during contacts selection

ubarsaiyan committed:
Fix ticks appearing on non selected contacts

ubarsaiyan committed:
Fix selected contacts not showing when selected through search and then coming back to the screen via back button

ubarsaiyan committed:
Remove invite contacts when creating groups

ear-dev commented:
@ubarsaiyan This looks good, I'll merge it.  Lets make the next priority to get the INVITE list back in if possible......

ubarsaiyan commented:
@ear-dev yes, sure :)
